sldns/str2wire.c

```diff
@@ -1511,7 +1516,70 @@ sldns_str2wire_svcbparam_alpn_value(const char* val,
 	sldns_write_uint16(rd + 2, dst_len);
 	memcpy(rd + 4, unescaped_dst, dst_len);
 	*rd_len = 4 + dst_len;
-	
+
+	return LDNS_WIREPARSE_ERR_OK;
+}
+
+static int
+sldns_str2wire_svcbparam_dohpath_value(const char* val,
+	uint8_t* rd, size_t* rd_len)
+{
+	size_t val_len;
+	char* open_bracket, * close_bracket;
+	const char* next_char;
+	uint8_t expr_found = 0;
+
+	/* RFC6570#section-2.1
+	 * "The characters outside of expressions in a URI Template string are
+	 * intended to be copied literally"
+	 * Practically this means we do not have to look for "double escapes"
+	 * like in the alpn value list.
+	 */
+
+	val_len = strlen(val);
+
+	if (*rd_len < 4 + val_len) {
+		return LDNS_WIREPARSE_ERR_BUFFER_TOO_SMALL;
+	}
+
+	/* draft-ietf-add-svcb-dns-06#section-5.1
+	 * "The URI Template MUST contain a "dns" variable"
+	 * A URI Template is alowed to have multiple variables
+	 */
+	next_char = val;
+	while (next_char && *next_char != '\0') {
+		char* c;
+
+		open_bracket = strchr(next_char, '{');
+		if (!open_bracket) {
+			return LDNS_WIREPARSE_ERR_SVCB_NO_DNS_VAR_IN_DOHPATH;
+			break;
+		}
+
+		close_bracket = strchr(open_bracket, '}');
+		if (!close_bracket) {
+			return LDNS_WIREPARSE_ERR_SVCB_NO_DNS_VAR_IN_DOHPATH;
+
+		}
+		for (c = open_bracket+1; (close_bracket - c) >= 4; c++) {
+			if (c[0] == '?' && c[1] == 'd' && c[2] == 'n'
+				&& c[3] == 's') {
+				expr_found++;
+			}
+		}
+
+		next_char = close_bracket+1;
+	}
+
+	if (expr_found != 1) {
+		return LDNS_WIREPARSE_ERR_SVCB_NO_DNS_VAR_IN_DOHPATH;
+	}
```

Comment:
This scares me a bit because it does not anticipate dohpaths in utf-8. The curly bracket could be part of a utf-8 character. WDYT?